build_defs/java_opts.bzl

```diff
@@ -26,8 +47,122 @@ def protobuf_java_library(**kwargs):
         **kwargs
     )
 
+def protobuf_java_module(
+    name,
+    module,
+    module_deps = [],
+    deps = [],
+    jvm_version = JPMS_JDK_TARGET,
+    visibility = DEFAULT_JMOD_VISIBILITY,
+    **kwargs):
+    """Builds a `module-info.java` definition for use with a Protobuf Java target.
+
+    This macro replaces a chain of `java_library` and `genrule` steps to coax Bazel
+    into placing the `module-info.class` at the right path in a versioned modular
+    MRJAR (Multi-Release JAR).
+
+    The name provided to this target is then meant to be used downstream as a
+    JAR-unwrapped classfile. For example:
```

Comment:
New macro for building a Java Module for use in the Protobuf Java library; example of use:

```starlark
protobuf_java_module(
    name = "my-module",
    module = "src/some/module-info.java",
)
java_library(
    name = "...",
    resources = [":my-module"],
    resource_strip_prefix = "src/some/",
)
```

The macro creates a target tree that looks like this:
```starlark
java_library(
  name = "module",
  srcs = [module],
  javacopts = JAVA9_OPTS,
  deps = deps_rewritten + (deps or []),  # rewritten module deps + regular deps
  visibility = visibility,
)
genrule(
  name = "module-class",
  srcs = [":module"],
  command = "mkdir -p META-INF/versions/9 && unzip module.jar module-info.class > META-INF/versions/9/module-info.class",
)
genrule(
  name = "module-class-zip",
  srcs = [":module-class"]
  command = "zip -r - META-INF > module-java9.jar",
)
java_import(
  name = "module-jar",
  jars = [":module-java9.jar"]
)

# This target is used in the `resources = [ ... ]` attr of a `java_library`
alias(
  name = "name-of-module-macro-target",
  actual = "module-class",
)

# This (hidden) target is used internally by the macro to target `module_deps`
alias(
  name = "name-of-module-macro-target-jar",
  actual = "module-jar",
)
```